tests/elasticarray/main.c

```diff
@@ -0,0 +1,167 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "elasticarray.h"
+
+#define NUM_ELEM 3
+
+struct stuff {
+	int i; 
+};
+
+struct dyn {
+	int * arr;
+};
+
+ELASTICARRAY_DECL(INTLIST, intlist, int);
+ELASTICARRAY_DECL(STUFFLIST, stufflist, struct stuff);
+ELASTICARRAY_DECL(DYNLIST, dynlist, struct dyn *);
+
+/* Elasticarray with simple integers. */
+static int
+test_intlist()
+{
+	INTLIST list;
+	size_t i;
+	int x;
+
+	/* Allocate list. */
+	if ((list = intlist_init(0)) == NULL)
+		goto err0;
+
+	/* Append some values. */
+	for (i = 0; i < NUM_ELEM; i++) {
+		x = (int)i;
+		if (intlist_append(list, &x, 1))
+			goto err0;
+	}
+
+	/* Use the list. */
+	for (i = 0; i < intlist_getsize(list); i++) {
+		x = *intlist_get(list, i);
+		printf("int:\t%i\n", x);
+	}
+
+	/* Free memory. */
+	intlist_free(list);
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Elasticarray with simple structures. */
+static int
+test_stufflist()
+{
+	STUFFLIST list;
+	struct stuff s;
+	size_t i;
+
+	/* Allocate list. */
+	if ((list = stufflist_init(0)) == NULL)
+		goto err0;
+
+	/* Append some values. */
+	for (i = 0; i < NUM_ELEM; i++) {
+		s.i = (int)i;
+
+		/* Append item. */
+		if (stufflist_append(list, &s, 1))
+			goto err0;
+	}
+
+	/* Use the list. */
+	for (i = 0; i < stufflist_getsize(list); i++) {
+		s = *stufflist_get(list, i);
+		printf("stuff:\t%i\n", s.i);
+	}
+
+	/* Free memory. */
+	stufflist_free(list);
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Elasticarray with dynamically-allocated structures. */
+static int
+test_dynlist()
+{
+	DYNLIST list;
+	struct dyn * p;
+	size_t i;
+
+	/* Allocate list. */
+	if ((list = dynlist_init(0)) == NULL)
+		goto err0;
+
+	/* Append some values. */
+	for (i = 0; i < NUM_ELEM; i++) {
+		/* Dynamically allocate item to append. */
+		if ((p = malloc(sizeof(struct dyn))) == NULL)
+			goto err0;
+		if ((p->arr = malloc(sizeof(int))) == NULL)
+			goto err0;
+		p->arr[0] = (int)i;
+
+		/* Append item. */
```

Comment:
I'd say "Append pointer to structure" here -- since `list` is an array of *pointers* rather than an array of structures

Reply:
Right, that leads me back to where I started getting into trouble: I wanted an array of structs, but dynamically allocated.

I think the missing part was to `free(p)` after appending, not at clean-up -- because the contents of `p` had been copied to the list.  See the `REBASE` commit.

BTW, would you recommend the current syntax of
```
struct dyn * p;
...
p = dynlist_get(list, i);
```
or do you think it's cleaner to do
```
struct dyn d;
struct dyn * p = &d;
...
d = *dynlist_get(list, i);
```

Reply:
Sometimes you want an elastic array of pointers.  Sometimes you want an elastic array of structures.  In this particular case, `struct dyn` has a fixed size so it's hard to see why you would want to malloc one, append (a copy of) it, and then free it.

As for `struct dyn * p = dynlist_get(list, i);` vs. `struct dyn d = *dynlist_get(list, i);` -- those do different things.  The former makes `p` point at the structure inside the elastic array; the latter copies the structure out of the array.